src/utilities.rs

```
@@ -50,14 +50,23 @@ pub fn get_ipv4_address(interface: &str) -> Option<Ipv4Addr> {
             _ => unreachable!(),
         })
         .collect();
+
     // Prefer Internet-routable addresses over loopback over private unicast.
     // TODO: The following line requires Rust nightly:
     // addresses.sort_by_key(|addr| (addr.is_global(), addr.is_loopback(), addr.is_private()));
+
+    // If the user has specified an address, use it if it is in the list of addresses.
```

Any reason to check that the user specified address is in the list?

Otherwise let's simplify this by just not calling `get_ipv4_address` if the user specified an address.

i.e. remove the `ipv4_src_addr` args and at the callsite do something like `ipv4_src_addr or get_ipv4_address(interface)`.

---

Agreed, retrospectively I'm not a huge fan of defaulting to another source address if Caracat does not find the inputed address on the interface. It's better to stick to what the user ask for, even though it can create a networking error.

---

Yeah; I prefer dropped packets because the IP is invalid/not routed, rather than silently defaulting to another address :)